File: auto_permissions/providers/openai_compat.py
"""OpenAI-compatible connector for llama.cpp / llama-server, vLLM, and LM Studio."""
import json
import time
import tempfile
import urllib.request
from pathlib import Path
from typing import Any, Dict, Optional

from .base import BaseProvider, parse_json_safely

class OpenAICompatibleProvider(BaseProvider):
    """Connects to OpenAI-compatible endpoints with auto model resolution, remote cloud auth, and dual-port fallback."""
    def __init__(
        self,
        endpoint: str = "http://127.0.0.1:9931/v1/chat/completions",
        model: str = "auto",
        api_key: Optional[str] = None,
        temperature: float = 0.0,
        timeout: float = 3.5,
    ):
        # Normalize localhost to 127.0.0.1 to avoid Windows IPv6/NetBIOS 1-2.5s DNS delays
        self.endpoint = endpoint.replace("localhost", "127.0.0.1")
        self.model = model
        self.api_key = api_key
        self.temperature = temperature
        self.timeout = timeout
        self._cached_model_id: Optional[str] = None

    def _resolve_model_id(self) -> str:
        """Resolve model name; auto-detects from /v1/models if local model is 'auto' or 'default'."""
        if self.endpoint.startswith("https://"):
            # Remote cloud endpoint (OpenAI, OpenRouter, Groq)
            return self.model if self.model not in ("auto", "default") else "gpt-4o-mini"

        if self.model and self.model not in ("auto", "default"):
            return self.model

        if self._cached_model_id:
            return self._cached_model_id

        # Check cross-process temp cache file with 60s TTL
        cache_file = Path(tempfile.gettempdir()) / "auto_permissions_model_llama.json"
        now = time.time()
        if cache_file.is_file():
            try:
                with open(cache_file, "r", encoding="utf-8") as f:
                    cache_data = json.load(f)
                if now - cache_data.get("cached_at", 0) < 60:
                    model_id = cache_data.get("model_id")
                    if model_id:
                        self._cached_model_id = model_id
                        return model_id
            except Exception:
                pass

        # Query /v1/models with strict 1.0s timeout (try primary endpoint first, then legacy 8080)
        candidate_endpoints = [self.endpoint]
        if ":9931" in self.endpoint:
            candidate_endpoints.append(self.endpoint.replace(":9931", ":8080"))

        for ep in candidate_endpoints:
            models_endpoint = ep.rsplit("/chat/completions", 1)[0] + "/models"
            try:
                req = urllib.request.Request(models_endpoint, headers={"Accept": "application/json"})
                with urllib.request.urlopen(req, timeout=1.0) as resp:
                    data = json.loads(resp.read().decode("utf-8"))
                    model_list = data.get("data") or data.get("models") or []
                    if model_list and isinstance(model_list, list):
                        detected = model_list[0].get("id") or model_list[0].get("name") or "default"
                        self._cached_model_id = detected
                        self.endpoint = ep  # Switch to active port
                        try:
                            with open(cache_file, "w", encoding="utf-8") as f:
                                json.dump({"model_id": detected, "cached_at": now}, f)
                        except Exception:
                            pass
                        return detected
            except Exception:
                continue

        return "default"

    def evaluate(self, system_prompt: str, prompt: str) -> Optional[Dict[str, Any]]:
        resolved_model = self._resolve_model_id()
        payload = {
            "model": resolved_model,
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": prompt}
            ],
            "temperature": self.temperature,
            "response_format": {"type": "json_object"},
        }
        data = json.dumps(payload).encode("utf-8")

        endpoints_to_try = [self.endpoint]
        if not self.endpoint.startswith("https://"):
            if ":9931" in self.endpoint and self.endpoint.replace(":9931", ":8080") not in endpoints_to_try:
                endpoints_to_try.append(self.endpoint.replace(":9931", ":8080"))

        headers = {"Content-Type": "application/json", "Accept": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"

        for ep in endpoints_to_try:
            req = urllib.request.Request(ep, data=data, headers=headers)
            try:
                with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                    resp_data = json.loads(resp.read().decode("utf-8"))
                    choices = resp_data.get("choices", [])
                    if choices:
                        content = choices[0].get("message", {}).get("content", "")
                        parsed = parse_json_safely(content)
                        if parsed:
                            self.endpoint = ep
                            return parsed
            except Exception:
                continue

        return None

File: internal/providers/openai_compat.go
package providers

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

// OpenAICompatibleProvider connects to llama-server, vLLM, LM Studio, Groq, OpenRouter, or OpenAI.
type OpenAICompatibleProvider struct {
	Endpoint        string
	Model           string
	APIKey          string
	Temperature     float64
	Timeout         time.Duration
	MaxTokens       int
	cachedModelID   string
	cachedModelTime time.Time
	mu              sync.Mutex
	client          *http.Client
}

func NewOpenAICompatibleProvider(endpoint, model, apiKey string, temp float64, timeout time.Duration, maxTokens int) *OpenAICompatibleProvider {
	if endpoint == "" {
		endpoint = "http://127.0.0.1:9931/v1/chat/completions"
	}
	endpoint = strings.ReplaceAll(endpoint, "localhost", "127.0.0.1")
	if maxTokens <= 0 {
		maxTokens = 512
	}
	if timeout <= 0 {
		timeout = 3500 * time.Millisecond
	}
	return &OpenAICompatibleProvider{
		Endpoint:    endpoint,
		Model:       model,
		APIKey:      apiKey,
		Temperature: temp,
		Timeout:     timeout,
		MaxTokens:   maxTokens,
		client:      &http.Client{Timeout: timeout},
	}
}

func (p *OpenAICompatibleProvider) GetEndpoint() string {
	return p.Endpoint
}

func (p *OpenAICompatibleProvider) ResolveModelID() string {
	p.mu.Lock()
	defer p.mu.Unlock()

	isRemote := strings.HasPrefix(p.Endpoint, "https://")
	if p.Model != "" && p.Model != "auto" && p.Model != "default" {
		return p.Model
	}

	if isRemote && strings.Contains(p.Endpoint, "api.openai.com") {
		return "gpt-4o-mini"
	}

	if p.cachedModelID != "" && time.Since(p.cachedModelTime) < 60*time.Second {
		return p.cachedModelID
	}

	cacheFile := filepath.Join(os.TempDir(), "auto_permissions_model_llama.json")
	if !isRemote {
		if data, err := os.ReadFile(cacheFile); err == nil {
			var cd struct {
				ModelID  string  `json:"model_id"`
				CachedAt float64 `json:"cached_at"`
			}
			if err := json.Unmarshal(data, &cd); err == nil {
				if float64(time.Now().Unix())-cd.CachedAt < 60 && cd.ModelID != "" {
					p.cachedModelID = cd.ModelID
					p.cachedModelTime = time.Now()
					return cd.ModelID
				}
			}
		}
	}

	candidateEndpoints := []string{p.Endpoint}
	if !isRemote && strings.Contains(p.Endpoint, ":9931") {
		candidateEndpoints = append(candidateEndpoints, strings.ReplaceAll(p.Endpoint, ":9931", ":8080"))
	}

	client := &http.Client{Timeout: 1000 * time.Millisecond}
	for _, ep := range candidateEndpoints {
		base := ep
		if idx := strings.LastIndex(base, "/chat/completions"); idx != -1 {
			base = base[:idx]
		}
		modelsURL := base + "/models"

		req, err := http.NewRequest("GET", modelsURL, nil)
		if err != nil {
			continue
		}
		req.Header.Set("Accept", "application/json")
		if p.APIKey != "" {
			req.Header.Set("Authorization", "Bearer "+p.APIKey)
		}

		resp, err := client.Do(req)
		if err != nil {
			continue
		}
		defer func() { _ = resp.Body.Close() }()

		if resp.StatusCode == http.StatusOK {
			body, _ := io.ReadAll(resp.Body)
			var respData map[string]interface{}
			if err := json.Unmarshal(body, &respData); err == nil {
				var list []interface{}
				if d, ok := respData["data"].([]interface{}); ok {
					list = d
				} else if m, ok := respData["models"].([]interface{}); ok {
					list = m
				}
				if len(list) > 0 {
					if first, ok := list[0].(map[string]interface{}); ok {
						name, _ := first["id"].(string)
						if name == "" {
							name, _ = first["name"].(string)
						}
						if name != "" {
							p.cachedModelID = name
							p.cachedModelTime = time.Now()
							if !isRemote {
								p.Endpoint = ep
								b, _ := json.Marshal(map[string]interface{}{
									"model_id":  name,
									"cached_at": float64(time.Now().Unix()),
								})
								_ = os.WriteFile(cacheFile, b, 0644)
							}
							return name
						}
					}
				}
			}
		}
	}

	if isRemote {
		return "gpt-4o-mini"
	}
	return "default"
}

func (p *OpenAICompatibleProvider) Evaluate(systemPrompt, prompt string) (map[string]interface{}, string, error) {
	resolvedModel := p.ResolveModelID()
	payload := map[string]interface{}{
		"model": resolvedModel,
		"messages": []map[string]string{
			{"role": "system", "content": systemPrompt},
			{"role": "user", "content": prompt},
		},
		"max_tokens":  p.MaxTokens,
		"temperature": p.Temperature,
		"response_format": map[string]string{
			"type": "json_object",
		},
		"chat_template_kwargs": map[string]bool{
			"enable_thinking": false,
		},
		"reasoning_effort": "none",
	}

	data, err := json.Marshal(payload)
	if err != nil {
		return nil, "", err
	}

	endpointsToTry := []string{p.Endpoint}
	if !strings.HasPrefix(p.Endpoint, "https://") && strings.Contains(p.Endpoint, ":9931") {
		fallback8080 := strings.ReplaceAll(p.Endpoint, ":9931", ":8080")
		if fallback8080 != p.Endpoint {
			endpointsToTry = append(endpointsToTry, fallback8080)
		}
	}

	source := "LOCAL"
	if strings.HasPrefix(p.Endpoint, "https://") {
		source = "CLOUD"
	}

	for _, ep := range endpointsToTry {
		req, err := http.NewRequest("POST", ep, bytes.NewReader(data))
		if err != nil {
			continue
		}
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("Accept", "application/json")
		if p.APIKey != "" {
			req.Header.Set("Authorization", "Bearer "+p.APIKey)
		}

		resp, err := p.client.Do(req)
		if err != nil {
			continue
		}
		defer func() { _ = resp.Body.Close() }()

		if resp.StatusCode == http.StatusOK {
			body, _ := io.ReadAll(resp.Body)
			var respData map[string]interface{}
			if err := json.Unmarshal(body, &respData); err == nil {
				if choices, ok := respData["choices"].([]interface{}); ok && len(choices) > 0 {
					if firstChoice, ok := choices[0].(map[string]interface{}); ok {
						if msg, ok := firstChoice["message"].(map[string]interface{}); ok {
							content, _ := msg["content"].(string)
							parsed := ParseJSONSafely(content)
							if parsed != nil {
								p.Endpoint = ep
								return parsed, source, nil
							}
						}
					}
				}
			}
		}
	}

	return nil, source, fmt.Errorf("openai compatible provider failed across endpoints: %v", endpointsToTry)
}
